src/sys/fuchsia/selector.rs

```diff
@@ -0,0 +1,370 @@
+    token_to_fd: Mutex<hash_map::HashMap<Token, Weak<EventedFdInner>>>,
+}
+
+impl Selector {
+    pub fn new() -> io::Result<Selector> {
+        let port = Arc::new(
+            magenta::Port::create(magenta::PortOpts::Default)
+                .map_err(status_to_io_err)?
+        );
+
+        // offset by 1 to avoid choosing 0 as the id of a selector
+        let id = NEXT_ID.fetch_add(1, Ordering::Relaxed) + 1;
+
+        let has_tokens_to_rereg = AtomicBool::new(false);
+        let tokens_to_rereg = Mutex::new(Vec::new());
+        let token_to_fd = Mutex::new(hash_map::HashMap::new());
+
+        Ok(Selector {
+            id,
+            port,
+            has_tokens_to_rereg,
+            tokens_to_rereg,
+            token_to_fd,
+        })
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    /// Returns a reference to the underlying port `Arc`.
+    pub fn port(&self) -> &Arc<magenta::Port> { &self.port }
+
+    /// Reregisters all registrations pointed to by the `tokens_to_rereg` list
+    /// if `has_tokens_to_rereg`.
+    fn reregister_handles(&self) -> io::Result<()> {
+        if self.has_tokens_to_rereg.load(Ordering::Relaxed) {
+            let mut tokens = self.tokens_to_rereg.lock().unwrap();
+            let token_to_fd = self.token_to_fd.lock().unwrap();
+            for token in tokens.drain(0..) {
+                if let Some(eventedfd) = token_to_fd.get(&token)
+                    .and_then(|h| h.upgrade()) {
+                    eventedfd.rereg_for_level(&self.port);
+                }
+            }
+            self.has_tokens_to_rereg.store(false, Ordering::Relaxed);
+        }
+        Ok(())
+    }
+
+    pub fn select(&self,
+                  evts: &mut Events,
+                  _awakener: Token,
+                  timeout: Option<Duration>) -> io::Result<bool>
+    {
+        evts.events.drain(0..);
+
+        self.reregister_handles()?;
+
+        let deadline = match timeout {
+            Some(duration) => {
+                let nanos = duration.as_secs().saturating_mul(1_000_000_000)
+                                .saturating_add(duration.subsec_nanos() as u64);
+
+                magenta::deadline_after(nanos)
+            }
+            None => magenta::MX_TIME_INFINITE,
+        };
+
+        let packet = match self.port.wait(deadline) {
+            Ok(packet) => packet,
+            Err(magenta::Status::ErrTimedOut) => return Ok(false),
+            Err(e) => return Err(status_to_io_err(e)),
+        };
+
+        let observed_signals = match packet.contents() {
+            magenta::PacketContents::SignalOne(signal_packet) => {
+                signal_packet.observed()
+            }
+            magenta::PacketContents::SignalRep(signal_packet) => {
+                signal_packet.observed()
+            }
+            magenta::PacketContents::User(_user_packet) => {
+                // User packets are only ever sent by an Awakener
+                return Ok(true);
```

Comment:
mio just cares that it was awakened by the `Awakener`-- it doesn't care about the token.